testsuite/MDAnalysisTests/analysis/test_density.py

```diff
@@ -105,6 +112,69 @@ def test_check_set_unit_nolength(self, D):
         with pytest.raises(ValueError):
             D._check_set_unit(units)
 
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        assert D._check_set_unit(units) == None
+
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        D._check_set_unit(units)
+        del D.units['density']
+        D._check_set_unit({})
+        assert D.units['density'] == None
+
+    def test_parameters_isdensity(self, D):
+        with pytest.warns(UserWarning, match='Running make_density()'):
+            D.make_density()
+
+    def test_check_convert_density_grid_not_density(self, D1):
+        unit = 'nm^{-3}'
+        with pytest.raises(RuntimeError, match="The grid is not a density"):
+            D1.convert_density(unit)
+
+    def test_check_convert_density_value_error(self, D):
+        unit = 'A^{-2}'
+        with pytest.raises(ValueError, match="The name of the unit"):
+            D.convert_density(unit)
+
+    def test_check_convert_density_units_same_density_units(self, D):
+        unit = 'A^{-3}'
+        D.convert_density(unit)
+        assert unit == D.units['density']
+
+    def test_check_convert_density_units_density(self, D):
+        unit = 'nm^{-3}'
+        D.convert_density(unit)
+        assert D.units['density'] == unit
+
+    def test_convert_length_same_length_units(self, D):
+        unit = 'A'
+        D.convert_length(unit)
+        assert unit == D.units['length']
+        unit='nm'
+        D.convert_length(unit)
+        assert unit == D.units['length']
+
+    def test_parameters_is_Density(self, D, D1):
+        D.__repr__()
+        D1.__repr__()
+        assert 'isDensity' in D.parameters
+        assert 'isDensity' in D1.parameters
+
+    def test_check_convert_length_edges(self, D):
+        D1 = D
+        new_unit = 'nm'
+        D.convert_length(new_unit)
+        for prev_edge, conv_edge in zip(D1.edges, D.edges):
+            assert_almost_equal(prev_edge, conv_edge)
+
+    def test_check_convert_density_edges(self, D):
+        D1 = D
```

Comment:
Please double check, but I'm pretty sure this is a shallow copy, so it won't work as a comparison if you're modifying D in place (same for the test above).

Reply:
Is this better? (`grid` is missing, though)

```
    def test_check_convert_length(self, D, h_and_edges, bins):
        midpoints = [0.5*(bin[:-1] + bin[1:]) for bin in bins]
        deltas = np.array([self.Lmax])/np.array(self.nbins)
        origin = [mp[0] for mp in midpoints]
        h, edges = h_and_edges
        new_unit = 'nm'
        D.convert_length(new_unit)

        for prev_edge, conv_edge in zip(edges, D.edges):
            assert_almost_equal(prev_edge*0.1,  conv_edge)
        
        for prev_origin, conv_origin in zip(origin, D.origin):
            assert_almost_equal(prev_origin*0.1,  conv_origin)
        
        for prev_delta, conv_delta in zip(deltas, D.delta):
            assert_almost_equal(prev_delta*0.1,  conv_delta)
```

Reply:
You can use `copy.deepcopy()` before changing the Density.

`Density` could really have a `Density.copy()` method to do it by itself, but that's for another day.